File: bundle-plugin/src/main/scala/com/yahoo/container/plugin/bundle/AnalyzeBundle.scala
// Copyright 2016 Yahoo Inc. Licensed under the terms of the Apache 2.0 license. See LICENSE in the project root.
package com.yahoo.container.plugin.bundle

import java.util.jar.{Manifest => JarManifest}
import java.io.File
import com.yahoo.container.plugin.osgi.{ExportPackages, ExportPackageParser}
import ExportPackages.Export
import collection.immutable.LinearSeq
import com.yahoo.container.plugin.util.JarFiles


/**
 * @author  tonytv
 */
object AnalyzeBundle {
  case class PublicPackages(exports : List[Export], globals : List[String])

  def publicPackagesAggregated(jarFiles : Iterable[File]) = aggregate(jarFiles map  {publicPackages(_)})

  def aggregate(publicPackagesList : Iterable[PublicPackages]) =
    (PublicPackages(List(), List()) /: publicPackagesList) { (a,b) =>
      PublicPackages(a.exports ++ b.exports, a.globals ++ b.globals)
    }

  def publicPackages(jarFile: File): PublicPackages = {
    try {

      (for {
        manifest <- JarFiles.getManifest(jarFile)
        if isOsgiManifest(manifest)
      } yield PublicPackages(parseExports(manifest), parseGlobals(manifest))).
      getOrElse(PublicPackages(List(), List()))

    } catch {
      case e : Exception => throw new RuntimeException("Invalid manifest in bundle '%s'".format(jarFile.getPath), e)
    }
  }

  def bundleSymbolicName(jarFile: File): Option[String] = {
    JarFiles.getManifest(jarFile).flatMap(getBundleSymbolicName)
  }

  private def parseExportsFromAttribute(manifest : JarManifest, attributeName : String) = {
    (for (export <- getMainAttributeValue(manifest, attributeName)) yield
      ExportPackageParser.parseAll(export) match {
        case noSuccess: ExportPackageParser.NoSuccess => throw new RuntimeException(
          "Failed parsing %s: %s".format(attributeName, noSuccess))
        case success => success.get
      }).
      getOrElse(List())
  }

  private def parseExports = parseExportsFromAttribute(_ : JarManifest, "Export-Package")

  private def parseGlobals(manifest : JarManifest) = {
    //TODO: Use separate parser for global packages.
    val globals = parseExportsFromAttribute(manifest, "Global-Package")

    if (globals map {_.parameters} exists {!_.isEmpty}) {
      throw new RuntimeException("Parameters not valid for Global-Package.")
    }

    globals flatMap {_.packageNames}
  }

  private def getMainAttributeValue(manifest: JarManifest, name: String): Option[String] =
    Option(manifest.getMainAttributes.getValue(name))

  private def isOsgiManifest = getBundleSymbolicName(_: JarManifest).isDefined

  private def getBundleSymbolicName = getMainAttributeValue(_: JarManifest, "Bundle-SymbolicName")
}

File: bundle-plugin/src/main/java/com/yahoo/container/plugin/bundle/AnalyzeBundle.java
// Copyright Yahoo. Licensed under the terms of the Apache 2.0 license. See LICENSE in the project root.
package com.yahoo.container.plugin.bundle;

import com.yahoo.container.plugin.osgi.ExportPackageParser;
import com.yahoo.container.plugin.osgi.ExportPackages.Export;
import com.yahoo.container.plugin.util.JarFiles;

import java.io.File;
import java.util.ArrayList;
import java.util.Collection;
import java.util.Collections;
import java.util.List;
import java.util.Optional;
import java.util.Set;
import java.util.jar.Manifest;

/**
 * Static utilities for analyzing jar files.
 *
 * @author Tony Vaagenes
 * @author ollivir
 * @author gjoranv
 */
public class AnalyzeBundle {

    public static List<Export> exportedPackagesAggregated(Collection<File> jarFiles) {
        List<Export> exports = new ArrayList<>();

        for (File jarFile : jarFiles) {
            var exported = exportedPackages(jarFile);
            exports.addAll(exported);
        }
        return exports;
    }

    static List<Export> exportedPackages(File jarFile) {
        try {
            Optional<Manifest> jarManifest = JarFiles.getManifest(jarFile);
            if (jarManifest.isPresent()) {
                Manifest manifest = jarManifest.get();
                if (isOsgiManifest(manifest)) {
                    return parseExports(manifest);
                }
            }
            return Collections.emptyList();
        } catch (Exception e) {
            throw new RuntimeException(String.format("Invalid manifest in bundle '%s'", jarFile.getPath()), e);
        }
    }

    public static Optional<String> bundleSymbolicName(File jarFile) {
        return JarFiles.getManifest(jarFile).flatMap(AnalyzeBundle::getBundleSymbolicName);
    }

    private static List<Export> parseExports(Manifest jarManifest) {
        return getMainAttributeValue(jarManifest, "Export-Package")
                .map(ExportPackageParser::parseExports)
                .orElseGet(ArrayList::new);
    }

    private static Optional<String> getMainAttributeValue(Manifest manifest, String attributeName) {
        return Optional.ofNullable(manifest.getMainAttributes().getValue(attributeName));
    }

    private static boolean isOsgiManifest(Manifest mf) {
        return getBundleSymbolicName(mf).isPresent();
    }

    private static Optional<String> getBundleSymbolicName(Manifest mf) {
        return getMainAttributeValue(mf, "Bundle-SymbolicName");
    }
}
